oauth2: Strict Matching

JediKev opened the pull request:
This is a rewrite of osTicket/osTicket-plugins#263 where it was attempting to implement Strict Matching. This fully implements Strict Matching and adds functionality to check for this when retrieving a token and when saving SMTP with Same As Remote Mailbox. This also adds a new disabled field to the General Settings to show what email will be used as well as a help tip that further explains Strict Matching.

JediKev committed:
oauth2: Strict Matching

This is a rewrite of osTicket/osTicket-plugins#263 where it was
attempting to implement Strict Matching. This fully implements Strict
Matching and adds functionality to check for this when retrieving a
token and when saving SMTP with Same As Remote Mailbox. This also adds a
new disabled field to the General Settings to show what email will be
used as well as a help tip that further explains Strict Matching.

protich commented on include/class.email.php:
```diff
@@ -487,6 +487,19 @@ public function isAuthBackendEnabled() {
     }
 
+    public function isStrict() {
+        $config = ($this instanceof SmtpAccount
```
Why not put this logic in `getConfig()` function?  

PS: We hadn't had an issue yet because getCredentials considered auth backend (i.e delegated to mailbox), but it makes sense to have it as part of getConfig.

protich commented on include/class.email.php:
```diff
@@ -487,6 +487,19 @@ public function isAuthBackendEnabled() {
+
+    public function checkStrictMatching($token=null) {
+        $token ??= $this->getAccessToken();
+        return $token->isMatch($this->getEmail()->getEmail(), $this->isStrict());
```
Token can be null. I would recommend.

```php
return ($token && $token->isMatch($this->getEmail()->getEmail(), $this->isStrict()))
```

MarcBanyard replied:
Just tested this and it needs to as per the recommendation from Peter as the token can be null.

JediKev replied:
Yes update coming soon

protich commented on include/class.email.php:
```diff
@@ -537,6 +553,11 @@ public function getEmail() {
 
+    public function getAccessToken() {
+        $cred = $this->getFreshCredentials();
+        return $cred->getAccessToken($this->getConfigSignature());
```
Credential can be null here!

```php
return ($cred && $cred->getAccessToken($this->getConfigSignature())); 
```

MarcBanyard replied:
Just tested this and it needs to as per the recommendation from Peter as the token can also be null.

protich reviewed (changes requested):
Some comments on the initial review.

protich commented:
To avoid requiring OAuth2 Plugin version upgrade in tandem with the core release - I'm suggesting we tag the strict mode flag to the backend id as shown below, for now, until we release v1.18.x series.

```diff
diff --git a/include/class.email.php b/include/class.email.php
index bc4943c6..0a1adea4 100644
--- a/include/class.email.php
+++ b/include/class.email.php
@@ -522,7 +522,8 @@ class EmailAccount extends VerySimpleModel {
             $id = sprintf('%s:%d',
                 $this->getAuthBk(), $this->getId());
             if ($this->isOAuthAuth())
-                $id .= sprintf(':%d', $this->getAuthId());
+                $id .= sprintf(':%d:%b',
+                    $this->getAuthId(), $this->isStrict());
 
             $this->bkId = $id;
         }
```

Thoughts? I'm open for a discussions as to why it makes sense as temp fix.

MarcBanyard commented:
> To avoid requiring OAuth2 Plugin version upgrade in tandem with the core release - I'm suggesting we tag the strict mode flag to the backend id as shown below, for now, until we release v1.18.x series.
> 
> ```diff
> diff --git a/include/class.email.php b/include/class.email.php
> index bc4943c6..0a1adea4 100644
> --- a/include/class.email.php
> +++ b/include/class.email.php
> @@ -522,7 +522,8 @@ class EmailAccount extends VerySimpleModel {
>              $id = sprintf('%s:%d',
>                  $this->getAuthBk(), $this->getId());
>              if ($this->isOAuthAuth())
> -                $id .= sprintf(':%d', $this->getAuthId());
> +                $id .= sprintf(':%d:%b',
> +                    $this->getAuthId(), $this->isStrict());
>  
>              $this->bkId = $id;
>          }
> ```
> 
> Thoughts? I'm open for a discussions as to why it makes sense as temp fix.

Why wait till v1.18.x series, as the fix for OAuth2 is required for users now.

I agree moving the **Strict Email Matching** to the core is a good idea, so the above pull (now with the recommended fixes for a null token) resolves the issue along with the minor change required in [osTicket/osTicket-plugins#265](https://github.com/osTicket/osTicket-plugins/pull/265) for the OAuth2 plugin fix the issue for everyone who uses Microsoft 365 with Shared Mailboxes.

The sooner this can be released to the public the better as it is very much needed since Microsoft forcefully disabled basic auth on all tenants.
Even if it comes out as v1.17.3 with a revised version of the plugin as per pull [osTicket/osTicket-plugins#265](https://github.com/osTicket/osTicket-plugins/pull/265) so users can update and get working using OAuth2 and ensure they do not access shared mailboxes with a username and password as this breaches Microsoft's licencing terms.

JediKev commented:
@MarcBanyard 

You are misinterpreting what he was saying. He's saying in order to avoid us having to make a database patch and force a major upgrade (1.18.x) we can simply add a temporary fix for now to release v1.17.4 and new builds of plugins (with the fixes included) and then worry about a major release later on.

Cheers.

JediKev committed:
oauth2: Add Strict To Backend ID

This adds the Strict Matching value to the Backend ID for now until we
add a column for it in the database (1.18.x).

MarcBanyard commented:
@JediKev & @protich any idea when you plan on releasing this as a temporary fix so the community can start using it?

JediKev committed:
oauth2: Delete Token

It's about damn time.. that we add ability to delete the OAuth2 Token.
This adds a new button to the Token tab that will allow you to delete
the current Token.

JediKev committed:
oauth2: Bug Fixes

This addresses some bugs found during testing. This adds a check to
`SmtpAccount::getConfig()` that ensures we have an email and
MailboxAccount before attempting to call `getConfig()` and adds
descriptive comments. This also removes the `isActive()` check from the
SmtpAccount `save()` method as we only care about if it's currently set to
active. Lastly, this changes the variable name for the email address
from `$email` to `$addr` so it doesn't conflict with the original
`$email` variable.

JediKev committed:
oauth2: Strict Matching Default True

This changes the default for Strict Matching from `false` to `true` so
that new emails are strict unless explicitly set otherwise. This also
adds some descriptive comments to the getter and setter.

JediKev committed:
oauth2: Delete Token Banner

This updates the Delete Token functionality to show a success/error
banner when deleting the token. If successful it will delete the Token
tab and content, load the IdP tab, and show a success banner with the
JSON content. On error it will simply show an error banner with the JSON
content.
